### PR TITLE
authz: protect instance endpoints

### PR DESCRIPTION
**Pull request by davepacheco**
This change adds protection for various Instance-related endpoints, following the same patterns as what's already been done for Disks and others.  This one's a bit more involved because of the start/stop/reboot endpoints.

### CODE REVIEW

**Comment by davepacheco** [comment]
I think this is now ready for review.

File: nexus/src/authz/api_resources.rs [hunk 1]
```diff
@@ -435,6 +435,10 @@ impl ProjectChild {
     pub fn id(&self) -> Uuid {
         self.resource_id
     }
+
+    pub fn project(&self) -> &Project {
+        &self.parent
+    }
```
Comment (david-crespo) [inline nexus/src/authz/api_resources.rs:441]: is this used?

Comment (davepacheco) [reply nexus/src/authz/api_resources.rs:441]: Yes, here:
https://github.com/oxidecomputer/omicron/blob/aa312ca5c18716c346828bf20a38b1c8680fda47/nexus/src/nexus.rs#L1557

File: nexus/tests/integration_tests/instances.rs [hunk 2]
```diff
@@ -238,13 +264,18 @@ async fn test_instances_create_reboot_halt(
...
+        StatusCode::BAD_REQUEST,
+        Method::POST,
+        &format!("{}/reboot", &instance_url),
+    )
+    .authn_as(AuthnMode::PrivilegedUser)
+    .execute()
+    .await
+    .unwrap()
+    .parsed_body()
+    .unwrap();
```
Comment (david-crespo) [inline nexus/tests/integration_tests/instances.rs:278]: I don't see an obvious way around it besides adding yet another helper, but these chains are long and there are a lot of them

File: nexus/src/authz/api_resources.rs (hunk 1 above)
Comment (david-crespo) [reply nexus/src/authz/api_resources.rs:441]: thanks. it didn't come up in my find because `nexus.rs` was collapsed

**Review by david-crespo — APPROVED** [review]
Looks good!

File: nexus/tests/integration_tests/instances.rs (hunk 2 above)
Comment (davepacheco) [reply nexus/tests/integration_tests/instances.rs:278]: I agree -- and a little error prone.  (It's easy to forget the `authn_as()` and get a 404 for the wrong reason.)  And you probably remember that the reason we _got_ here was we had like 5-10 helper functions that were all only slightly different from each other :-/

It's possible most of our tests should be using clients generated from the OpenAPI spec.  That could make a lot of these a lot clearer.

I've also been considering using something like [hurl](https://hurl.dev/) for our tests, where we use something like a DSL to describe the requests we want to make and the expected responses.  This sounds fantastic at first but it gets hairy when you add authn and stuff like the side-door operations used by the tests to manipulate Nexus state outside the API.

**Pull request merged**